File: analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Mutex.cs
using System;
using System.Threading;

namespace Mutex_Type
{
    // https://docs.microsoft.com/en-us/dotnet/standard/threading/mutexes
    internal class Foo
    {
        bool cond;
        public Mutex instanceMutex;
        public static Mutex staticMutex;

        public void Noncompliant(Foo foo)
        {
            var m0 = new Mutex(true, "bar", out var m0WasCreated); // Noncompliant

            var m1 = new Mutex(false);
            m1.WaitOne(); // Noncompliant

            var m2 = new Mutex(false, "qix", out var m2WasCreated);
            m2.WaitOne(); // Noncompliant

            var m3 = Mutex.OpenExisting("x");
            m3.WaitOne(); // Noncompliant

            foo.instanceMutex.WaitOne(); // FN

            Foo.staticMutex.WaitOne(); // Noncompliant

            if (cond)
            {
                m0.ReleaseMutex();
                m1.ReleaseMutex();
                m2.ReleaseMutex();
                m3.ReleaseMutex();
                foo.instanceMutex.ReleaseMutex();
                Foo.staticMutex.ReleaseMutex();
            }

            // Note that Dispose() closes the underlying WaitHandle, but does not release the mutex
            m0.Dispose();
            m1.Dispose();
            m2.Dispose();
            m3.Dispose();
        }

        public void MutexWasCreatedAndReleased(bool arg)
        {
            var m = new Mutex(initiallyOwned: true, "bar", out var wasCreated); // Noncompliant
            if (wasCreated && arg)
            {
                m.ReleaseMutex();
            }
        }

        public void MutexWasCreatedAndNotReleased(bool arg)
        {
            var m = new Mutex(initiallyOwned: true, "bar", out var wasCreated); // Compliant
            if (!wasCreated && arg)
            {
                // wasCreated is false here, indicating that the lock is NOT (yet) held
                m.ReleaseMutex(); // This is a release without a previous lock. This is not covered by this rule.
            }
        }

        public void MutexWasCreatedWithoutOwnership(bool arg)
        {
            var m = new Mutex(initiallyOwned: false, "bar", out var wasCreated); // Compliant. The lock must still be acquired by requesting it.
            if (wasCreated && arg)
            {
                m.ReleaseMutex();
            }
        }

        public void Noncompliant2(Mutex paramMutex, Mutex paramMutex2)
        {
            // 'true' means it owns the mutex if no exception gets thrown
            using (var mutexInUsing = new Mutex(true, "foo")) // Noncompliant
            {
                if (cond)
                {
                    mutexInUsing.ReleaseMutex();
                }
            }

            if (Mutex.TryOpenExisting("y", out var mutexInOutVar))
            {
                mutexInOutVar.WaitOne(); // Noncompliant
                if (cond)
                {
                    mutexInOutVar.ReleaseMutex();
                }
            }

            var m = new Mutex(false);
            var mIsAcquired = m.WaitOne(200, true);
            if (mIsAcquired)
            {
                // here it should be released
            }
            else
            {
                m.ReleaseMutex(); // this is a programming error, but not subject of this rule
            }

            var paramMutexIsAcquired = paramMutex.WaitOne(400, false); // Noncompliant
            if (paramMutexIsAcquired)
            {
                if (cond)
                {

                }
                else
                {
                    paramMutex.ReleaseMutex();
                }
            }
            while (paramMutex2.WaitOne(400, false)) // Noncompliant
            {
                if (cond)
                {
                    paramMutex2.ReleaseMutex();
                }
            }
        }

        public void NoncompliantReleasedThenAcquiredAndReleased(Mutex paramMutex)
        {
            paramMutex.ReleaseMutex();
            paramMutex.WaitOne(); // Noncompliant
            if (cond)
            {
                paramMutex.ReleaseMutex();
            }
        }

        public void DifferentInstancesOnThis(Foo foo)
        {
            foo.instanceMutex.WaitOne(); // Compliant
            instanceMutex.WaitOne(); // Noncompliant
            if (cond)
            {
                instanceMutex.ReleaseMutex();
            }
        }

        public void DifferentInstancesOnParameter(Foo foo)
        {
            foo.instanceMutex.WaitOne(); // FN
            instanceMutex.WaitOne(); // Compliant
            if (cond)
            {
                foo.instanceMutex.ReleaseMutex();
            }
        }

        public void UnsupportedWaitAny(Mutex m1, Mutex m2, Mutex m3)
        {
            // it is too complex to support this scenario
            WaitHandle[] handles = new[] { m1, m2, m3 };
            var index = WaitHandle.WaitAny(handles); // FN
                                                     // the mutex at the given index should be released
            var acquiredMutex = (Mutex)handles[index];
            if (cond)
            {
                acquiredMutex.ReleaseMutex();
            }
        }

        public void UnsupportedWaitAll(Mutex m1, Mutex m2, Mutex m3)
        {
            // it is too complex to support this scenario
            WaitHandle[] handles = new[] { m1, m2, m3 };
            var allHaveBeenAcquired = WaitHandle.WaitAll(handles); // FN
            if (allHaveBeenAcquired)
            {
                // all indexes should be released
                if (cond)
                {
                    ((Mutex)handles[0]).ReleaseMutex();
                }
            }
        }

        public void CompliantAcquiredNotReleased(Mutex paramMutex, Foo foo)
        {
            using (var m = new Mutex(true, "foo"))
            {
                // do stuff
            }

            var m0 = new Mutex(true, "bar", out var m0WasCreated);
            m0.Dispose();

            var m1 = new Mutex(false);
            m1.WaitOne();
            m1.Dispose();

            var m2 = new Mutex(false, "qix", out var m2WasCreated);
            m2.WaitOne();
            m2.Dispose();

            var m3 = Mutex.OpenExisting("x");
            m3.WaitOne();

            if (Mutex.TryOpenExisting("y", out var m4))
            {
                m4.WaitOne();
            }

            var isAcquired = paramMutex.WaitOne(400, false);
            if (isAcquired)
            {
                // not released
            }

            foo.instanceMutex.WaitOne();
            Foo.staticMutex.WaitOne();
        }

        public void CompliantNotAcquired(Mutex paramMutex)
        {
            var m1 = new Mutex(false);
            var m2 = Mutex.OpenExisting("foo");
            if (Mutex.TryOpenExisting("foo", out var m3))
            {
                // do stuff but don't acquire
            }
            var m4 = new Mutex(false, "foo", out var mutexWasCreated);
            if (paramMutex != null)
            {
                // do stuff but don't acquire
            }
        }

        public void CompliantAcquiredAndReleased(Mutex paramMutex, Foo foo)
        {
            var m1 = new Mutex(false);
            m1.WaitOne();
            m1.ReleaseMutex();

            var m2 = Mutex.OpenExisting("foo");
            if (m2.WaitOne(500))
            {
                m2.ReleaseMutex();
            }

            var isAcquired = paramMutex.WaitOne(400, false);
            if (isAcquired)
            {
                paramMutex.ReleaseMutex();
            }
            if (paramMutex.WaitOne(400, false))
            {
                paramMutex.ReleaseMutex();
            }

            foo.instanceMutex.WaitOne();
            if (cond)
            {
                foo.instanceMutex.ReleaseMutex();
            }
            else
            {
                foo.instanceMutex.ReleaseMutex();
            }

            while (cond)
            {
                Foo.staticMutex.WaitOne();
                Foo.staticMutex.ReleaseMutex();
            }
        }

        public void ReleasedThenAcquired(Mutex paramMutex)
        {
            if (cond)
            {
                paramMutex.ReleaseMutex();
            }
            paramMutex.WaitOne(); // Compliant, source of FPs
        }

        public void CompliantComplex(string mutexName, bool shouldAcquire)
        {
            Mutex m = null;
            bool acquired = false;
            try
            {
                m = Mutex.OpenExisting(mutexName);
                if (shouldAcquire)
                {
                    m.WaitOne();    // Compliant, depends on tracking Null constraint for 'm'
                    acquired = true;
                }
            }
            catch (UnauthorizedAccessException)
            {
                return;
            }
            finally
            {
                if (m != null)
                {
                    // can enter also if an exception was thrown when Waiting
                    if (acquired)
                    {
                        m.ReleaseMutex();
                    }
                    m.Dispose();
                }
            }
        }

        public void MutextAquireByConstructor_SimpleAssignment_LiteralArgument()
        {
            var m = new Mutex(true, "bar", out var mutextCreated); // Noncompliant
            if (cond)
            {
                m.ReleaseMutex();
            }
        }

        public void MutextAquireByConstructor_SimpleAssignment_Array()
        {
            Mutex[] arr = new Mutex[2];
            arr[0] = new Mutex(true, "bar", out var mutextCreated); // FN, untracked symbol
            if (cond)
            {
                arr[0].ReleaseMutex();
            }
        }

        public void MutextAquireByConstructor_SimpleAssignment_FieldArgument()
        {
            var m = new Mutex(cond, "bar", out var mutextCreated);
            if (cond)
            {
                m.ReleaseMutex();
            }
        }

        public void MutextAquireByConstructor_ReAssignment()
        {
            var m = new Mutex(false, "bar", out var mc1);
            m = new Mutex(true, "bar", out var mc2); // Noncompliant
            if (cond)
            {
                m.ReleaseMutex();
            }
        }

        public void MutextAquireByConstructor_MultipleVariableDeclaration()
        {
            Mutex m0, m1;
            m0 = m1 = new Mutex(true, "bar", out var mutextCreated); // FN

            if (cond)
            {
                m0.ReleaseMutex();
            }
        }

        public void MutextAquireByConstructor_SwitchExpression(int x)
        {
            var m = x switch
            {
                1 => new Mutex(),
                2 => new Mutex(new bool()),
                3 => new Mutex(true, "bar", out var mutextCreated), // FN
            };

            if (cond)
            {
                m.ReleaseMutex();
            }
        }

        public void NullTrackingCondition()
        {
            if (instanceMutex != null)
            {
                instanceMutex.WaitOne();    // Compliant, depends on tracking null conditions
            }


            if (instanceMutex != null)
            {
                instanceMutex.ReleaseMutex();
            }
        }
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Mutex.vb
Imports System.Threading

Namespace Mutex_Type

    '    https://docs.microsoft.com/en-us/dotnet/standard/threading/mutexes
    Class Foo

        Private Cond As Boolean
        Public InstanceMutex As Mutex
        Public Shared StaticMutex As Mutex

        Public Sub Noncompliant(Foo As Foo)
            Dim m0WasCreated, m2WasCreated As Boolean
            Dim m0 = New Mutex(True, "bar", m0WasCreated) ' Noncompliant

            Dim m1 = New Mutex(False)
            m1.WaitOne() ' Noncompliant

            Dim m2 = New Mutex(False, "qix", m2WasCreated)
            m2.WaitOne() ' Noncompliant

            Dim m3 = Mutex.OpenExisting("x")
            m3.WaitOne() ' Noncompliant

            Foo.InstanceMutex.WaitOne() ' FN

            Foo.StaticMutex.WaitOne() ' Noncompliant

            If Cond Then
                m0.ReleaseMutex()
                m1.ReleaseMutex()
                m2.ReleaseMutex()
                m3.ReleaseMutex()
                Foo.InstanceMutex.ReleaseMutex()
                Foo.StaticMutex.ReleaseMutex()
            End If

            ' Note that Dispose() closes the underlying WaitHandle, but does Not release the mutex
            m0.Dispose()
            m1.Dispose()
            m2.Dispose()
            m3.Dispose()
        End Sub

        Public Sub Noncompliant2(ParamMutex As Mutex, ParamMutex2 As Mutex)

            ' 'true' means it owns the mutex if no exception gets thrown
            Using mutexInUsing As New Mutex(True, "foo") ' Noncompliant
                If Cond Then mutexInUsing.ReleaseMutex()
            End Using

            Dim mutexInOutVar As Mutex
            If Mutex.TryOpenExisting("y", mutexInOutVar) Then
                mutexInOutVar.WaitOne() ' Noncompliant
                If Cond Then mutexInOutVar.ReleaseMutex()
            End If

            Dim m = New Mutex(False)
            Dim mIsAcquired = m.WaitOne(200, True)
            If mIsAcquired Then
                ' here it should be released
            Else
                m.ReleaseMutex() ' This is a programming mistake, not detected by this rule
            End If

            Dim paramMutexIsAcquired = ParamMutex.WaitOne(400, False) ' Noncompliant
            If paramMutexIsAcquired Then
                If Cond Then
                Else
                    ParamMutex.ReleaseMutex()
                End If
            End If
            While ParamMutex2.WaitOne(400, False) ' Noncompliant
                If Cond Then ParamMutex2.ReleaseMutex()
            End While
        End Sub

        Public Sub NoncompliantReleasedThenAcquiredAndReleased(ParamMutex As Mutex)
            ParamMutex.ReleaseMutex()
            ParamMutex.WaitOne() ' Noncompliant
            If Cond Then ParamMutex.ReleaseMutex()
        End Sub

        Public Sub DifferentInstancesOnThis(Foo As Foo)
            Foo.InstanceMutex.WaitOne() ' Compliant
            InstanceMutex.WaitOne() ' Noncompliant
            If Cond Then InstanceMutex.ReleaseMutex()
        End Sub

        Public Sub DifferentInstancesOnParameter(Foo As Foo)
            Foo.InstanceMutex.WaitOne() ' FN
            InstanceMutex.WaitOne() ' Compliant
            If Cond Then Foo.InstanceMutex.ReleaseMutex()
        End Sub

        Public Sub UnsupportedWaitAny(m1 As Mutex, m2 As Mutex, m3 As Mutex)
            ' it Is too complex to support this scenario
            Dim WHandles() As WaitHandle = {m1, m2, m3}
            Dim Index = WaitHandle.WaitAny(WHandles) ' FN
            ' the mutex at the given index should be released
            Dim acquiredMutex = DirectCast(WHandles(Index), Mutex)
            If Cond Then acquiredMutex.ReleaseMutex()
        End Sub

        Public Sub UnsupportedWaitAll(m1 As Mutex, m2 As Mutex, m3 As Mutex)
            ' it Is too complex to support this scenario
            Dim WHandles() As WaitHandle = {m1, m2, m3}
            Dim allHaveBeenAcquired = WaitHandle.WaitAll(WHandles) ' FN
            If allHaveBeenAcquired Then
                ' all indexes should be released
                If Cond Then DirectCast(WHandles(0), Mutex).ReleaseMutex()
            End If
        End Sub

        Public Sub CompliantAcquiredNotReleased(ParamMutex As Mutex, Foo As Foo)
            Dim m0WasCreated, m2WasCreated As Boolean, m4 As Mutex
            Using m As New Mutex(True, "foo")
                ' do stuff
            End Using

            Dim m0 = New Mutex(True, "bar", m0WasCreated)
            m0.Dispose()

            Dim m1 = New Mutex(False)
            m1.WaitOne()
            m1.Dispose()

            Dim m2 = New Mutex(False, "qix", m2WasCreated)
            m2.WaitOne()
            m2.Dispose()

            Dim m3 = Mutex.OpenExisting("x")
            m3.WaitOne()

            If Mutex.TryOpenExisting("y", m4) Then m4.WaitOne()

            Dim isAcquired = ParamMutex.WaitOne(400, False)
            If isAcquired Then
                ' Not released
            End If

            Foo.InstanceMutex.WaitOne()
            Foo.StaticMutex.WaitOne()
        End Sub

        Public Sub CompliantNotAcquired(ParamMutex As Mutex)
            Dim m1 As New Mutex(False)
            Dim m2 As Mutex = Mutex.OpenExisting("foo")
            Dim m3 As Mutex
            If Mutex.TryOpenExisting("foo", m3) Then
                ' do stuff but don't acquire
            End If
            Dim MutexWasCreated As Boolean
            Dim m4 As New Mutex(False, "foo", MutexWasCreated)
            If ParamMutex IsNot Nothing Then
                ' do stuff but don't acquire
            End If
        End Sub

        Public Sub CompliantAcquiredAndReleased(ParamMutex As Mutex, Foo As Foo)
            Dim m1 As New Mutex(False)
            m1.WaitOne()
            m1.ReleaseMutex()

            Dim m2 = Mutex.OpenExisting("foo")
            If m2.WaitOne(500) Then
                m2.ReleaseMutex()
            End If

            Dim isAcquired = ParamMutex.WaitOne(400, False)
            If isAcquired Then ParamMutex.ReleaseMutex()

            If ParamMutex.WaitOne(400, False) Then
                ParamMutex.ReleaseMutex()
            End If

            Foo.InstanceMutex.WaitOne()
            If Cond Then
                Foo.InstanceMutex.ReleaseMutex()
            Else
                Foo.InstanceMutex.ReleaseMutex()
            End If

            While Cond
                Foo.StaticMutex.WaitOne()
                Foo.StaticMutex.ReleaseMutex()
            End While
        End Sub

        Public Sub ReleasedThenAcquired(ParamMutex As Mutex)
            ParamMutex.WaitOne() ' Noncompliant
            If Cond Then ParamMutex.ReleaseMutex()
        End Sub

        Public Sub CompliantComplex(MutexName As String, ShouldAcquire As Boolean)
            Dim M As Mutex, Acquired As Boolean = False
            Try
                M = Mutex.OpenExisting(MutexName)
                If ShouldAcquire Then
                    M.WaitOne()     ' Compliant, depends on tracking Null constraint for 'm'
                    Acquired = True
                End If
            Catch ex As UnauthorizedAccessException
                Return
            Finally
                If M Is Nothing Then
                    ' can enter also if an exception was thrown when Waiting
                    If Acquired Then M.ReleaseMutex()
                    M.Dispose()
                End If
            End Try
        End Sub

        Public Sub MutexAquireByConstructor_SimpleAssignment_LiteralArgument()
            Dim MutexCreated As Boolean
            Dim m = New Mutex(True, "bar", MutexCreated) ' Noncompliant
            If Cond Then m.ReleaseMutex()
        End Sub

        Public Sub MutexAquireByConstructor_SimpleAssignment_FieldArgument()
            Dim MutexCreated As Boolean
            Dim m = New Mutex(Cond, "bar", MutexCreated)
            If Cond Then m.ReleaseMutex()
        End Sub

        Public Sub MutexAquireByConstructor_ReAssignment()
            Dim mc1, mc2 As Boolean
            Dim m As New Mutex(False, "bar", mc1)
            m = New Mutex(True, "bar", mc2) ' Noncompliant
            If Cond Then m.ReleaseMutex()
        End Sub

        Public Sub MutexAquireByConstructor_MultipleVariableDeclaration()
            Dim MutexCreated As Boolean
            Dim m0, m1 As New Mutex(True, "bar", MutexCreated) ' Noncompliant
            If Cond Then m0.ReleaseMutex()
        End Sub

    End Class

End Namespace
